### DISCUSSION TITLE
Turning Python objects that Uproot would write to ROOT files directly into their PyROOT equivalents

### DISCUSSION THREAD

**Discussion by Esword618** [scikit-hep/uproot5#1151] [Q&A]
Here's my code.
I want to make the code from `# start` to `# end` simpler, is there any functional interface for this? I didn't find it.
```
c11 = TCanvas("c11", "c11",1000,600)
# start
temF = uproot.recreate("tem.root")
temF["inieKine"] = hist11
h1f11 = uproot.to_pyroot(temF["inieKine"])
# end
h1f11.GetXaxis().SetRangeUser(0,2e+6)
h1f11.SetTitle("inieKine")
h1f11.SetXTitle("Mev")
h1f11.SetYTitle("Count")
h1f11.Draw()
c11.Draw()
```

**Comment by jpivarski** [comment]
I see that you want to make a ROOT TH1F from a NumPy array, but don't want to have to write it to a file to do so. (It's not just about simplicity; it's about the fact that disk is orders of magnitude slower than RAM.)

The function that is called to create an Uproot TH1F object is [uproot.writing.identify.to_TH1x](https://uproot.readthedocs.io/en/latest/uproot.writing.identify.to_TH1x.html). It has a lot of arguments, so that the TH1F can be fully specified (ROOT histograms have a lot of metadata). Note that you can use the [uproot.writing.identify.to_TAxis](https://uproot.readthedocs.io/en/latest/uproot.writing.identify.to_TAxis.html) function to make the `fAxis` argument. The Uproot TH1F object can be converted into a PyROOT TH1F object with [uproot.to_pyroot](https://uproot.readthedocs.io/en/latest/uproot.pyroot.to_pyroot.html). This serializes and deserializes the data, since Uproot's memory model (Python) is completely unlike PyROOT's memory model (C++), but at least the serialized data doesn't go through disk.

That function has a lot of arguments, though. The function that is called to turn anything that you're assigning into a file is [uproot.writing.identify.to_writable](https://uproot.readthedocs.io/en/latest/uproot.writing.identify.to_writable.html). Since `hist11` is something that Uproot recognizes and knows to convert into a TH1F, you can replace the code between `# start` and `# end` with

```python
h1f11 = uproot.to_pyroot(uproot.writing.identify.to_writable(hist11))
```

While I'm at it, though, I should point out that [hist](https://hist.readthedocs.io/) histograms have plotting capabilities in the Matplotlib world, in case you're open to either Matplotlib or ROOT Canvases. It's hard to mix those two worlds, but if you didn't know about hist plotting and the rest of your analysis plotting is in Matplotlib, you should know that this is an option.

Since this isn't really a request for a new feature—the functions you specifically asked about exist—I'm going to turn this into a Discussion, so that it serves as a "how to" document.

**Reply by Esword618** [reply]
I know Matplotlib can draw images, but I find the ones drawn by ROOT to be much more attractive and convenient. At present, I use uproot to process data (numba is used for large data), and then use ROOT for graph drawing, which I feel is very convenient, because pyroot graphics in jupyter lab are interactive, it is very convenient to see some data, so no Matplotlib for drawing.
